api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java

```diff
@@ -13,30 +15,67 @@
  */
 public interface AsyncStreamingCache<K> {
    /**
-    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure
-    * that the stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its
-    * raw binary form. The returned input stream implements the {@link CacheEntryMetadata} interface.
-    * The returned input stream is not thread-safe.
+    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure that the
+    * stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its raw binary form. The
+    * returned input stream implements the {@link CacheEntryMetadata} interface. The returned input stream is not
+    * thread-safe.
+    *
+    * @param key key to use
+    */
+   default <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key) {
+      return get(key, CacheOperationOptions.DEFAULT);
+   }
+
+   /**
+    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure that the
+    * stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its raw binary form. The
+    * returned input stream implements the {@link CacheEntryMetadata} interface. The returned input stream is not
+    * thread-safe.
     *
     * @param key      key to use
+    * @param metadata
+    */
+   <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key, CacheOperationOptions metadata);
+
+   /**
+    * Initiates a streaming put operation. It is up to the application to write to the returned {@link OutputStream} and
+    * close it when there is no more data to write. The marshaller is ignored, i.e. all data will be written in its raw
+    * binary form. The returned output stream is not thread-safe.
+    *
+    * @param key key to use
     */
-   <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key);
+   default CompletionStage<OutputStream> put(K key) {
```

Comment:
Emitter or something similar should work. However, looking at `MutinyStreamingCache` this is still going to be error prone for the non blocking case as the user should not keep emitting in a loop for example since there is no way to signal back pressure.

We may want to investigate using something similar to http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/core/Flowable.html#generate-io.reactivex.rxjava3.functions.Consumer-

Basically we would invoke the Consumer each time they can emit a new value and each invocation should only ever emit one entry.

Reply:
@wburns For Async I've decided to go with an interface which mimics the JDK HTTP client API. For Mutiny, @cescoffier may provide more insight.

Reply:
Sleeping on it the generate method doesn't work as well since the user has no way to signal when a new value is to be emitted as the generate method requires it to be available when the Consumer is invoked.

Looking at the JDK client API I see https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpResponse.BodySubscriber.html which makes sense.

I would say we should have the MutinyCache do similar to the AsyncCache.

Reply:
Done